### ISSUE TITLE
".extend()" and ".vstack()" raises "PanicException: The column lengths in the DataFrame are not equal" with dataframes of  same number of columns

### ISSUE THREAD

**Issue by DataEnggNerd** [pola-rs/polars#18765]
### Checks

- [X] I have checked that this issue has not already been reported.
- [X] I have confirmed this bug exists on the [latest version](https://pypi.org/project/polars/) of Polars.

### Reproducible example

```python
mongo_data.shape
# (17788485, 8)
req_data.shape
# (19590895, 8)
mongo_data.schema
# Schema([('district_code', String), ('taluka_code', String), ('village_code', String), ('old_survey_number', String), ('block_number_inner', String), ('survey_number', Int32), ('sub_survey_number', String), ('thandapper_number', String)])
req_data.schema
# Schema([('district_code', String), ('taluka_code', String), ('village_code', String), ('old_survey_number', String), ('block_number_inner', String), ('survey_number', Int32), ('sub_survey_number', String), ('thandapper_number', String)])
req_data.vstack(mongo_data)
# thread '<unnamed>' panicked at crates/polars-core/src/fmt.rs:540:13:
# The column lengths in the DataFrame are not equal.
# *** pyo3_runtime.PanicException: The column lengths in the DataFrame are not equal.
req_data.extend(mongo_data)
# thread '<unnamed>' panicked at crates/polars-core/src/fmt.rs:540:13:
# The column lengths in the DataFrame are not equal.
# *** pyo3_runtime.PanicException: The column lengths in the DataFrame are not equal.

```

### Issue description

I have two data frames which I want to combine. Both the data frames are having same number of columns, but vary by row count. When I try to either `.extend()` or `.vstack()`, I am getting the error mentioned above.

To check, I have took only top 10k rows and tried to apply `.extend()` or `.vstack()`, which worked. 

### Expected behavior

Dataframes should get merged as they are of same column shape. 

### Installed versions

<details>
--------Version info---------
Polars:              1.6.0
Index type:          UInt32
Platform:            macOS-14.6.1-arm64-arm-64bit
Python:              3.11.10 (main, Sep  7 2024, 01:03:31) [Clang 15.0.0 (clang-1500.3.9.4)]

----Optional dependencies----
adbc_driver_manager  <not installed>
altair               <not installed>
cloudpickle          3.0.0
connectorx           0.3.3
deltalake            <not installed>
fastexcel            <not installed>
fsspec               2024.6.1
gevent               <not installed>
great_tables         <not installed>
matplotlib           <not installed>
nest_asyncio         1.6.0
numpy                2.0.0
openpyxl             <not installed>
pandas               2.2.2
pyarrow              16.0.0
pydantic             2.7.1
pyiceberg            <not installed>
sqlalchemy           2.0.31
torch                <not installed>
xlsx2csv             <not installed>
xlsxwriter           <not installed>
</details>

**Comment by cmdlineluser** [comment]
It may be worth checking if `pl.concat` also errors.

(I think the devs will need a reproducible example in order to be able to fix this.)

**Comment by RobLW** [comment]
It's a tricky bug to track down. I'm no programmer but I'm guessing this is limited to Jupyter Notebooks (iPython) only as it requires code to be run in different cells to re-create the error.

In Jypyter notebook, place this in a cell
```Python
import polars as pl
columns = {'Code': pl.UInt16, 'ContentCORRECT': pl.String, 'GroupName': pl.String}
df1 = pl.DataFrame([],schema=columns)
```
Now put this into a 2nd cell. If you place all code into a single cell, the issue doesn't exist.
```python
df2 = pl.DataFrame(
    data={'Code':[1,2],'AlienType':['z','y']}, 
    schema={'Code':pl.UInt16,'AlienType':pl.String}
)
df2 = df2.with_columns(GroupName=pl.lit(df2.columns[1])) # Add GroupName column and fill it with name of 2nd column
df2 = df2.rename({df2.columns[1]:'ContentWRONG'}) # Rename condition column
#df2 = df2.rename({df2.columns[1]:'ContentCORRECT'}) # Rename condition column

df1.extend(df2)
```

When you run it all you will get a very helpful error to inform you that your column names do not match. 
In **cell2** uncomment the line which contains _ContentCORRECT_, and comment out the line above it, then run the cell again to see the error.
I tried to attached the ipynb file as a demo for you, but seems this isn't allowed.

A reset of the python kernel should get most people quickly passed the problem until an update is out, as they would have corrected the column names before restarting the kernel.

HTH

**Comment by pj-ml** [comment]
Perhaps this will be helpful in narrowing down the bug:
```python
import polars as pl
df = pl.DataFrame({"x": range(10)})
df2 = pl.DataFrame({"x": range(10, 20)})
df = df.vstack(df2)
# does not error

agg_df = pl.DataFrame()
for i in range(1):
    df = pl.DataFrame({"x": range(10)})
    agg_df = agg_df.vstack(df)
# does not error


agg_df = pl.DataFrame()
for i in range(2):
    df = pl.DataFrame({"x": range(10)})
    agg_df = agg_df.vstack(df)
# errors with the following:
# "PanicException: The column lengths in the DataFrame are not equal."
```

**Comment by pj-ml** [comment]
I would guess this is related to #19641

**Comment by pj-ml** [comment]
Last python version of polars where this was still working was `1.9.0`

**Comment by jasheldo** [comment]
> It may be worth checking if `pl.concat` also errors.
> 
> (I think the devs will need a reproducible example in order to be able to fix this.)

I can confirm this is a bug for `pl.concat`.

I'm working on making a reproducible example.

**Comment by jasheldo** [comment]
Reproducible Example (Python 3.12.7, macOS Sequoia 15.1, M1)

```{python}
df = pl.DataFrame()
df1 = pl.DataFrame([{'year': 2024, 'region': 'f3snacktown', 'date_awarded': datetime.date(2024, 11, 9), 'pax_id': 'MY_AWESOME_ID'}])

pl.concat(
	[
		df, df1.with_columns(pl.lit(11).alias("achievement_id")).select(
			"achievement_id", "pax_id", "date_awarded"
		),
	]
)
```

yields StackTrace:

```
thread '<unnamed>' panicked at crates/polars-core/src/fmt.rs:567:13:
The column lengths in the DataFrame are not equal.
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/USER/Library/Caches/pypoetry/virtualenvs/venv-93dzw48B-py3.12/lib/python3.12/site-packages/polars/dataframe/frame.py", line 1181, in __repr__
    return self.__str__()
           ^^^^^^^^^^^^^^
  File "/Users/USER/Library/Caches/pypoetry/virtualenvs/venv-93dzw48B-py3.12/lib/python3.12/site-packages/polars/dataframe/frame.py", line 1178, in __str__
    return self._df.as_str()
           ^^^^^^^^^^^^^^^^^
pyo3_runtime.PanicException: The column lengths in the DataFrame are not equal.
```

**Comment by cmdlineluser** [comment]
@jasheldo The empty DataFrame case should be fixed on 1.13.0 (just released)

```python
>>> pl.__version__
'1.13.0'
```
```
shape: (1, 3)
┌────────────────┬───────────────┬──────────────┐
│ achievement_id ┆ pax_id        ┆ date_awarded │
│ ---            ┆ ---           ┆ ---          │
│ i32            ┆ str           ┆ date         │
╞════════════════╪═══════════════╪══════════════╡
│ 11             ┆ MY_AWESOME_ID ┆ 2024-11-09   │
└────────────────┴───────────────┴──────────────┘
```

I think the original example in the issue may be different as there are no empty frames involved - but it's impossible to say without a repro.

**Comment by DataEnggNerd** [comment]
Thank you for the active contribution and the fixes. I am closing the issue.
